assets/js/modules/analytics-4/components/widgets/TopTrafficSourceWidget.js

```diff
@@ -21,27 +21,133 @@
  */
 import PropTypes from 'prop-types';
 
+/**
+ * WordPress dependencies
+ */
+import { __, sprintf } from '@wordpress/i18n';
+
 /**
  * Internal dependencies
  */
 import Data from 'googlesitekit-data';
+import MetricTileText from '../../../../components/KeyMetrics/MetricTileText';
 import { CORE_USER } from '../../../../googlesitekit/datastore/user/constants';
+import {
+	DATE_RANGE_OFFSET,
+	MODULES_ANALYTICS_4,
+} from '../../datastore/constants';
+import { numFmt } from '../../../../util';
 
-const { useSelect } = Data;
+const { useSelect, useInViewSelect } = Data;
 
 export default function TopTrafficSourceWidget( { Widget, WidgetNull } ) {
 	const keyMetricsWidgetHidden = useSelect( ( select ) =>
 		select( CORE_USER ).isKeyMetricsWidgetHidden()
 	);
 
+	const dates = useSelect( ( select ) =>
+		select( CORE_USER ).getDateRangeDates( {
+			offsetDays: DATE_RANGE_OFFSET,
+			compare: true,
+		} )
+	);
+
+	const totalUsersReportOptions = {
+		...dates,
+		metrics: [
+			{
+				name: 'totalUsers',
+			},
+		],
+	};
+
+	const trafficSourceReportOptions = {
+		...dates,
+		dimensions: [ 'sessionDefaultChannelGroup' ],
+		metrics: [
+			{
+				name: 'totalUsers',
+			},
+		],
+		limit: 1,
+		orderBy: 'totalUsers',
+	};
+
+	const totalUsersReport = useInViewSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getReport( totalUsersReportOptions )
+	);
+
+	const trafficSourceReport = useInViewSelect( ( select ) =>
+		select( MODULES_ANALYTICS_4 ).getReport( trafficSourceReportOptions )
+	);
+
+	const loading = useSelect(
+		( select ) =>
+			! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
+				'getReport',
+				[ totalUsersReportOptions ]
+			) ||
+			! select( MODULES_ANALYTICS_4 ).hasFinishedResolution(
+				'getReport',
+				[ trafficSourceReportOptions ]
+			)
+	);
+
+	const totalUsers =
+		parseInt(
+			totalUsersReport?.rows?.[ 1 ]?.metricValues[ 0 ]?.value,
+			10
+		) || 0;
+
+	const topTrafficSource =
+		trafficSourceReport?.rows?.[ 1 ]?.dimensionValues[ 0 ]?.value || '-';
+
+	const currentTopTrafficSourceUsers =
+		parseInt(
+			trafficSourceReport?.rows?.[ 1 ]?.metricValues[ 0 ]?.value,
+			10
+		) || 0;
+	const relativeCurrentTopTrafficSourceUsers = totalUsers
+		? currentTopTrafficSourceUsers / totalUsers
+		: 0;
+
+	const previousTopTrafficSourceUsers =
+		parseInt(
+			trafficSourceReport?.rows?.[ 0 ]?.metricValues[ 0 ]?.value,
+			10
+		) || 0;
```

Comment:
It's quite hard to understand why `1` or `0` rows is selected for one or another value. And if something changes the ordering of those rows, everything will be immediately broken.

Could you, please, use similar approach to the one we use in LoyalVisitors and NewVisitors widgets where we define which row we want to find based on the dimension value(s) and then take metric value? https://github.com/google/site-kit-wp/blob/c1024b3e156d370c167067fc58710946c4efdb44/assets/js/modules/analytics-4/components/widgets/LoyalVisitorsWidget.js#L74-L94 https://github.com/google/site-kit-wp/blob/c1024b3e156d370c167067fc58710946c4efdb44/assets/js/modules/analytics-4/components/widgets/NewVisitorsWidget.js#L74-L90